docs/guides/00-overview/02-community/01-rfcs/20231227-fine-grained-storage-security.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Fine Grained Storage Security
+description: Add fine grained storage security to replace allow_insecure
+---
+
+- RFC PR: [datafuselabs/databend-docs#291](https://github.com/datafuselabs/databend-docs/pull/291)
+- Tracking Issue: [datafuselabs/databend#0000](https://github.com/datafuselabs/databend/issues/0000)
+
+## Summary
+
+Add fine grained storage security to replace `allow_insecure`.
+
+## Motivation
+
+Databend previously disabled all insecure storage services by default, preventing the use of `fs` or `s3` services over the HTTP protocol. This measure protects our users from attempts to steal data or probe for internal network services. For instance, using an SQL query like `SELECT * FROM 'http://192.168.100.100:8000'` could potentially expose sensitive information and serve as a gateway to more significant attacks.
+
+We support only a basic switch named `allow_insecure`. This is overly restrictive for users looking to access data from public HTTP services, yet too permissive for those who wish to avoid reading from the local filesystem. We need a more fine-grained approach to storage security.
+
+## Guide-level explanation
+
+Users can now have more control over storage security via `[storage.security]`:
+
+```toml
+[storage.security]
```

Comment:
Maybe whitelist mode is more widely used?

Reply:
> Maybe whitelist mode is more widely used?

Maybe we can support both of them in the future.

At current stage, I feel like disallow list is good enough for most of our usage. For example, our cloud users may want to `select * from http://example.com`. The domain list could be very large that we can't write in whitelist mode.

Reply:
There also could be an option like `sslmode`.

Reply:
> There also could be an option like `sslmode`.

The `allow_insecure` option functions similarly to `sslmode`, though it's not particularly useful. Users are looking to query HTTP and Minio services. Once we restrict the domain and CIDR, is `sslmode` still necessary? I mean, it should be safe for cloud to query like `select * from http://my.http.website.com`?

Reply:
Plain http should not be allowed on cloud, since it could lead to security issues.

Reply:
> Plain http should not be allowed on cloud, since it could lead to security issues.

Would you like to elaborate further? This RFC aims to tackle these issues, enabling users to query plain HTTP data sources.

Reply:
For example, people could set dns record for an external domain to an internal address.

Reply:
> Plain http should not be allowed on cloud, since it could lead to security issues.

Another option is adding `allowed_xxx` now, and cloud can use the allowed style. Users can submit ticket to add trusted domain.

Reply:
> For example, people could set dns record for an external domain to an internal address.

Oh, that's ok. We will resovle the domain first and check `blocked_cidr` before sending requests.

Reply:
Hi, @everpcpc, I have added `allowed_*`, please review again.

Reply:
> Oh, that's ok. We will resovle the domain first and check `blocked_cidr` before sending requests. The only thing we need to do is adding our internal address into `blocked_cidr`.

That's just an example, plain http request on public network could result in hijacking or other attacks and should be avoid.

Reply:
> That's just an example, plain http request on public network could result in hijacking or other attacks and should be avoid.

Understood. By including `allowed_*`, we can restrict domains exclusively for trusted users, eliminating the need for an additional `sslmode`. What are your thoughts?

Reply:
I think domain whitelist and ssl mode are different things, and both are needed for cloud or opensource users.